accepted/tail-calls.md

```diff
@@ -0,0 +1,1239 @@
+  host-to-host call that doesn't need trampolines. This adds complexity and
+  overhead to our call paths. We want these to be as maintainable as possible
+  and definitely as fast as possible.
+
+* Add Cranelift as a build dependency for Wasmtime and pre-compile a trampoline
+  for each of our `IntoFunc` signatures so we always have trampolines available
+  for host functions. In addition to complicating the build process, this bloats
+  the binary with a bunch of trampolines that are probably never used, which
+  makes it harder to use Wasmtime in resource constrained environments.
+
+* Use the type system to differentiate between `wasmtime::HostFunc` and
+  `wasmtime::WasmFunc` and then statically disallow calls that would need some
+  sort of trampoline that we wouldn't have access to. This is a large, clunky
+  change to Wasmtime's API for very little benefit, especially since we have
+  viable alternatives.
+
+As far as the new Wasm calling convention goes, there are some alternative
+approaches that we could explore as well:
+
+* We could bend over backwards to maintain interoperability with existing
+  calling conventions (to a degree) the way that [SpiderMonkey plans on
+  doing](https://docs.google.com/document/d/19Q5yMe7GUwi8T7ynWtCcmmd1DPk9xz2GQVOInh_NR3U/edit).
+  This involves making the first tail call a regular call and then only
+  subsequent tail calls in the chain reuse stack frames. This is subtle and
+  requires dynamically checking whether a tail call is the first or not, which
+  our proposed design avoids. It does make sense for SpiderMonkey, though, where
+  there are more tiers of compilers that all need to agree with each other and
+  with the host. They have a lot more inertia than we do.
+
+* We could leave stack arguments before the return address and saved frame
+  pointer, and still have callees clean up the stack arguments. The previous
+  alternative design sort of implies this behavior, but it is worth exploring on
+  its own as well. The main problem here is that eventually the callee epilogue
+  is trying to clean up those stack arguments, but can't just pop them because
+  the return address is on top of them:
+
+  ```
+  [ ...caller-frame... | ...stack-args... | RA ]
+  ```
+
+  In general, this can be solved by copying the return address around, but this
+  has questionable interactions with hardware control-flow integrity features as
+  well as pessimizing speculation. Epilogues could also start to get lengthy, in
+  terms of how many instructions they contain, rather than remaining short,
+  tight sequences. x86 does actually have a form of `ret` where you can pass an
+  immediate that says how many additional bytes to pop from the stack after
+  popping the return address, and so it could avoid this particular downside,
+  but not all architectures have an equivalent. And why try to solve this
+  incidental problem when we have a viable alternative that side steps it
+  completely?
```

Comment:
This layout is actually better than the one proposed in this RFC because it avoids the issue at the call site where data is written below the stack pointer. Even though wasmtime itself uses an alternate signal stack, the embedder could also have set up another signal handler without `SA_ONSTACK` which would not use the alternate signal stack.

No copying of the return address is actually necessary since x86 is the only architecture which combines a pop of the return address in the same instruction as the return itself. Other architectures use a return address register which can be popped first, after which the stack arguments can be freed before actually returning.